worldedit-core/src/main/java/com/sk89q/worldedit/function/block/SnowSimulator.java

```diff
@@ -120,20 +120,19 @@ public boolean apply(BlockVector3 position, int depth) throws WorldEditException
                 abovePosition) > 10) {
             return false;
         } else if (!block.getBlockType().getMaterial().isFullCube()) {
-            Map<Property<?>, Object> states = block.getStates();
-            if (states.containsKey(slab) && block.getState(slab).equalsIgnoreCase("bottom")) {
+            BlockType type = block.getBlockType();
+            if (type.hasProperty(slab) && block.getState(slab).equalsIgnoreCase("bottom")) {
                 return false;
-            } else if (states.containsKey(trapdoorOpen) && states.containsKey(trapdoor) && (block.getState(trapdoorOpen)
-                    || block.getState(trapdoor).equalsIgnoreCase("bottom"))) {
+            } else if ((type.hasProperty(trapdoorOpen) && block.getState(trapdoorOpen)) ||
+                    (type.hasProperty(trapdoor) && block.getState(trapdoor).equalsIgnoreCase("bottom"))) {
                 return false;
-            } else if (states.containsKey(stair) && block.getState(stair).equalsIgnoreCase("bottom")) {
+            } else if (type.hasProperty(stair) && block.getState(stair).equalsIgnoreCase("bottom")) {
                 return false;
             } else {
                 return false;
             }
             //FAWE end
-        } else if (!block.getBlockType().id().toLowerCase(Locale.ROOT).contains("ice") && block
-                .getBlockType()
+        } else if (!block.getBlockType().id().toLowerCase(Locale.ROOT).contains("ice") && block.getBlockType()
```

Comment:
Worth updating maybe, minecraft provides block tags (present in the BlockCategories class as constants) `snow_layer_cannot_survive_on` and `snow_layer_can_survive_on`, which could make this more readable and future-proof (new ice blocks, new spawn logic, etc)

Reply:
Ah good shout, cheers